File: leetcode/366. Find Leaves of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        dic = {}
        
        def go_depth(node):
            if node is None:
                return 0
            
            depth = max(go_depth(node.right), go_depth(node.left)) + 1
            if depth not in dic:
                dic[depth] = []
            dic[depth].append(node.val)
        
            return depth
        
        depth = go_depth(root)
        
        return [dic[i] for i in range(1, depth+1)]

File: leetcode/366. Find Leaves of Binary Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;
use std::cmp;

impl Solution {
    pub fn find_leaves(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut map: HashMap<u8, Vec<i32>> = HashMap::new();
        
        fn go_depth(node: &Option<Rc<RefCell<TreeNode>>>, map: &mut HashMap<u8, Vec<i32>>) -> u8 {
            match node {
                Some(node_p) => {
                    let node = node_p.borrow();
                    let depth = cmp::max(
                        go_depth(&node.left, map),
                        go_depth(&node.right, map)
                    ) + 1;
                    
                    (*map.entry(depth).or_insert(Vec::new())).push(node.val);
                    return depth;
                },
                None => {
                    return 0;
                }
            }
        }
        
        let depth = go_depth(&root, &mut map);
        
        let mut res = Vec::new();
        for d in (1..depth+1) {
            res.push(map.remove(&d).unwrap());
        }
        
        res
    }
}
